bin/dockerd-wrapper

```diff
@@ -51,7 +51,7 @@ export XDG_RUNTIME_DIR=$SNAP_COMMON/run
 # use SNAP_DATA for most "data" bits
 mkdir -p \
 	"$SNAP_DATA/run" \
-	"$SNAP_DATA/run/docker" \
+	"/run/snap.$SNAP_INSTANCE_NAME" \
```

Comment:
Ah, interesting!  Is this an official path that'd documented somewhere for our use?
(Shouldn't there be another variable like `$SNAP_DATA` to point to it?  I don't see anything related to `/run` in https://snapcraft.io/docs/environment-variables :sweat_smile:)

If we switch to this, I don't think we need to create `"$SNAP_DATA/run"` at all, since `--exec-root` and `--pidfile` are all we use it for AFAIR, and it's much better (matches upstream's expectations) to put all that on a tmpfs. :smile:

Reply:
@anonymouse64 I don't recall seeing this documented, either.. I've tested the feature works, so I think the docs just haven't caught up with this feature yet...

Reply:
I see it defined in the common AppArmor profile at https://github.com/snapcore/snapd/blob/79aced6888fae0a5785c784d4aa4dba28992c236/interfaces/apparmor/template.go#L444-L447, so perhaps that's the closest thing we have to official "documentation" of it?

Reply:
(also in the packaging changelog at https://github.com/snapcore/snapd/blob/79aced6888fae0a5785c784d4aa4dba28992c236/packaging/ubuntu-16.04/changelog#L3766, but that's probably just sourced from the commit/PR title)

Reply:
It's not documented, but it should be :wink: , and there's not an env variable, but there should be :wink: 

see: https://github.com/snapcore/snapd/blob/79aced6888fae0a5785c784d4aa4dba28992c236/interfaces/apparmor/template.go#L444-L447

Reply:
@degville, we need this in a document please :-)

Reply:
> @degville, we need this in a document please :-)

will do! thanks for letting me know.

Reply:
So, I'd like to adjust our implementation of it here a little bit for now so we have a better chance of remembering. :sweat_smile:

Here's what I've got in mind:

```bash
# modify XDG_RUNTIME_DIR to be a snap writable dir underneath $SNAP_COMMON 
# (until https://bugs.launchpad.net/snappy/+bug/1656340 is fixed)
export XDG_RUNTIME_DIR="$SNAP_COMMON/run"
mkdir -p "$XDG_RUNTIME_DIR"

# use SNAP_COMMON for "/var/lib/docker" since it's common across snap revisions
# (and typically gets large, so we don't want it backed up and restored like SNAP_DATA)
mkdir -p "$SNAP_COMMON/var-lib-docker"

# use "/run/snap.$SNAP_INSTANCE_NAME" for ephemeral storage
# - https://github.com/snapcore/snapd/pull/6109
# - https://github.com/docker-snap/docker-snap/pull/9
run_dir="/run/snap.$SNAP_INSTANCE_NAME"
mkdir -p "$run_dir"

workaround_lp1606510

workaround_apparmor_profile_reload

exec dockerd \
	--group "$default_socket_group" \
	--exec-root="$run_dir" \
	--data-root="$SNAP_COMMON/var-lib-docker" \
	--pidfile="$run_dir/docker.pid" \
	--config-file="$SNAP_DATA/config/daemon.json" \
	"$@"
```

(Happy to take over and update/carry this PR directly if you'd prefer :+1:)

Reply:
@degville I'm by no means any kind of authority (just a humble nobody here :sweat_smile:), but it might also be worth adding a note about `XDG_RUNTIME_DIR` (which is set) that includes a caveat about https://bugs.launchpad.net/snappy/+bug/1656340 (which makes the provided value harder to use) :sweat_smile:

Reply:
@tianon, I've added 8cc8d9c to include your suggestions